tests/fv/node_auto_hep_test.go

```diff
@@ -0,0 +1,336 @@
+		Eventually(func() error {
+			_, err := k8sClient.CoreV1().Namespaces().List(metav1.ListOptions{})
+			return err
+		}, 30*time.Second, 1*time.Second).Should(BeNil())
+
+		// Run controller manager.  Empirically it can take around 10s until the
+		// controller manager is ready to create default service accounts, even
+		// when the hyperkube image has already been downloaded to run the API
+		// server.  We use Eventually to allow for possible delay when doing
+		// initial pod creation below.
+		controllerManager = testutils.RunK8sControllerManager(apiserver.IP)
+	})
+
+	AfterEach(func() {
+		os.Remove(kconfigFile.Name())
+		controllerManager.Stop()
+		nodeController.Stop()
+		apiserver.Stop()
+		etcd.Stop()
+	})
+
+	It("should create and sync hostendpoints for Calico nodes", func() {
+		// Run controller with auto HEP enabled
+		nodeController = testutils.RunNodeController(apiconfig.EtcdV3, etcd.IP, kconfigFile.Name(), true)
+
+		// Create a kubernetes node with some labels.
+		kn := &v1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: kNodeName,
+				Labels: map[string]string{
+					"label1": "value1",
+				},
+			},
+		}
+		_, err := k8sClient.CoreV1().Nodes().Create(kn)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Create a Calico node with a reference to it.
+		cn := calicoNode(c, cNodeName, kNodeName, map[string]string{"calico-label": "calico-value", "label1": "badvalue"})
+		_, err = c.Nodes().Create(context.Background(), cn, options.SetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		// Expect the node label to sync.
+		expected := map[string]string{"label1": "value1", "calico-label": "calico-value"}
+		Eventually(func() error { return testutils.ExpectNodeLabels(c, expected, cNodeName) },
+			time.Second*15, 500*time.Millisecond).Should(BeNil())
+
+		expectedHepName := cn.Name + "-auto-hep"
+
+		// Expect a wildcard hostendpoint to be created.
+		expectedLabels := map[string]string{
+			"label1":                       "value1",
+			"calico-label":                 "calico-value",
+			"projectcalico.org/created-by": "calico-kube-controllers",
+		}
+		expectedIPs := []string{"172.16.1.1", "fe80::1", "192.168.100.1"}
+		Eventually(func() error { return testutils.ExpectHostendpoint(c, expectedHepName, expectedLabels, expectedIPs) },
+			time.Second*15, 500*time.Millisecond).Should(BeNil())
+
+		// Update the Kubernetes node labels.
+		kn, err = k8sClient.CoreV1().Nodes().Get(kn.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		kn.Labels["label1"] = "value2"
+		_, err = k8sClient.CoreV1().Nodes().Update(kn)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Expect the node labels to sync.
+		expected = map[string]string{"label1": "value2", "calico-label": "calico-value"}
+		Eventually(func() error { return testutils.ExpectNodeLabels(c, expected, cNodeName) },
+			time.Second*15, 500*time.Millisecond).Should(BeNil())
+
+		// Expect the hostendpoint labels to sync.
+		expectedLabels["label1"] = "value2"
```

Comment:
Reading this, it's hard to distinguish `expected` vs `expectedLabels`. I think these should be called `expectedNodeLabels` and `expectedHEPLabels`, respectively?

Reply:
aaa55e056523cfd838c0767732bb248ae4f0ce80